### PR TITLE
Added additional DSL types and component

### PR DESCRIPTION
**Pull request by mrigankmg**
Added WithApplicability, WithPlayerTypes, and ObjectWithIndexTracking as part of https://github.com/player-ui/player/pull/367

### Change Type (required)
Indicate the type of change your pull request is:

- [x] `patch`
- [ ] `minor`
- [ ] `major`

### CODE REVIEW

File: language/dsl/src/components.tsx [hunk 1]
```diff
@@ -262,3 +262,13 @@ export function createSlot<SlotProps = unknown>(options: {
     );
   };
 }
+/** An object wrapper that tracks it's index from a parent */
+export const ObjectWithIndexTracking = (props: WithChildren) => {
```
Comment (KetanReddy) [inline language/dsl/src/components.tsx:266]: Can we add a test for this?

Comment (mrigankmg) [reply language/dsl/src/components.tsx:266]: What would testing something look like?

Comment (KetanReddy) [reply language/dsl/src/components.tsx:266]: Something similar to how we test the auto id functionality should be sufficient https://github.com/player-ui/tools/blob/main/language/dsl/src/__tests__/asset-api.test.tsx

File: language/dsl/src/__tests__/helpers/asset-library.tsx [hunk 2]
```diff
@@ -369,3 +375,113 @@ Info.AdditionalInfo = createSlot<{
...
+  return (
+    <Asset type="choice" {...rest}>
+      <property name="binding">{binding.toValue()}</property>
+      {children}
+    </Asset>
+  );
+};
+
+Choice.Title = createSlot<{
+  /** Some thing not in the asset  */
```
Comment (rafbcampos) [inline language/dsl/src/__tests__/helpers/asset-library.tsx:426]: Do we leverage that property (as we do on the `Input.Label`, for example)?

**Review by rafbcampos — APPROVED** [review]
I commented on the Slot creation, but it's not a blocker.

File: language/dsl/src/__tests__/helpers/asset-library.tsx (hunk 2 above)
Comment (mrigankmg) [reply language/dsl/src/__tests__/helpers/asset-library.tsx:426]: Could you clarify what you mean by "leverage that property"?

Comment (rafbcampos) [reply language/dsl/src/__tests__/helpers/asset-library.tsx:426]: You are extending all the Choice's Slots props with `customLabelProp`. We do that for `Input.Label` to later be able to do that:
```tsx
<Input.Label customLabelProp="custom label slot value">
```
But it doesn't seem it's needed to Choice.Note, Choice.Items etc.

**Review by rafbcampos — APPROVED** [review]

**Pull request merged**